File: retroauth-android/src/test/java/com/andretietz/retroauth/AccountAuthenticatorTest.java
package com.andretietz.retroauth;

import android.accounts.Account;
import android.accounts.AccountAuthenticatorResponse;
import android.accounts.AccountManager;
import android.accounts.NetworkErrorException;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import static com.andretietz.retroauth.testhelper.Helper.getMember;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.mockito.Mockito.mock;

@RunWith(RobolectricTestRunner.class)
public class AccountAuthenticatorTest {

    private AccountAuthenticator authenticator;

    @Before
    public void setup() {
        ActivityManager.Companion.get(RuntimeEnvironment.application);
        authenticator = new AccountAuthenticator(mock(Context.class), "some-action");
    }

    @Test
    public void constructor() throws NoSuchFieldException, IllegalAccessException {
        AccountAuthenticator authenticator = new AccountAuthenticator(mock(Context.class), "some-action");
        String action = getMember(authenticator, "action", String.class);
        assertNotNull(action);
        assertEquals("some-action", action);
    }

    @Test
    public void addAccount() throws NetworkErrorException {
        AccountAuthenticatorResponse response = mock(AccountAuthenticatorResponse.class);
        Bundle bundle = authenticator.addAccount(response, "accountType", "tokenType",
                new String[]{}, mock(Bundle.class));

        assertNotNull(bundle);
        Intent intent = bundle.getParcelable(AccountManager.KEY_INTENT);
        assertNotNull(intent);

        assertEquals(
                response,
                intent.getParcelableExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE));
        assertEquals("accountType", intent.getStringExtra(AccountManager.KEY_ACCOUNT_TYPE));
        assertEquals("tokenType", intent.getStringExtra(AccountAuthenticator.Companion.getKEY_TOKEN_TYPE()));
    }
    @Test
    public void getAuthToken() throws NetworkErrorException {
        AccountAuthenticatorResponse response = mock(AccountAuthenticatorResponse.class);
        Account account = new Account("accountName", "accountType");
        Bundle bundle = authenticator.getAuthToken(response, account, "tokenType", mock(Bundle.class));

        assertNotNull(bundle);
        Intent intent = bundle.getParcelable(AccountManager.KEY_INTENT);
        assertNotNull(intent);

        assertEquals(
                response,
                intent.getParcelableExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE));
        assertEquals("accountType", intent.getStringExtra(AccountManager.KEY_ACCOUNT_TYPE));
        assertEquals("tokenType", intent.getStringExtra(AccountAuthenticator.Companion.getKEY_TOKEN_TYPE()));
        assertEquals("accountName", intent.getStringExtra(AccountManager.KEY_ACCOUNT_NAME));
    }

    @Test
    public void hasFeatures() throws NetworkErrorException {
        Bundle bundle = authenticator
                .hasFeatures(mock(AccountAuthenticatorResponse.class), mock(Account.class), new String[]{});
        assertNull(bundle);
    }

    @Test
    public void updateCredentials() throws NetworkErrorException {
        Bundle bundle = authenticator
                .updateCredentials(mock(AccountAuthenticatorResponse.class), mock(Account.class), "token-type",
                        mock(Bundle.class));
        assertNull(bundle);
    }

    @Test
    public void getAuthTokenLabel() throws NetworkErrorException {
        String label = authenticator.getAuthTokenLabel("token-type");
        assertNull(label);
    }

    @Test
    public void editProperties() throws NetworkErrorException {
        Bundle bundle = authenticator
                .editProperties(mock(AccountAuthenticatorResponse.class), "accountType");
        assertNull(bundle);
    }

    @Test
    public void confirmCredentials() throws NetworkErrorException {
        Bundle bundle = authenticator
                .confirmCredentials(mock(AccountAuthenticatorResponse.class), mock(Account.class), mock(Bundle.class));
        assertNull(bundle);
    }
}

File: retroauth-android/src/test/java/com/andretietz/retroauth/AccountAuthenticatorTest.kt
package com.andretietz.retroauth

import android.accounts.Account
import android.accounts.AccountAuthenticatorResponse
import android.accounts.AccountManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.mock
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment

@RunWith(RobolectricTestRunner::class)
class AccountAuthenticatorTest {

  private val authenticator: AccountAuthenticator = AccountAuthenticator(mock(Context::class.java), "some-action")

  @Before
  fun setup() {
    ActivityManager[RuntimeEnvironment.application]
  }

  @Test
  fun constructor() {
    val authenticator = AccountAuthenticator(mock(Context::class.java), "some-action")
    assertNotNull(authenticator.action)
    assertEquals("some-action", authenticator.action)
  }

  @Test
  fun addAccount() {
    val response = mock(AccountAuthenticatorResponse::class.java)
    val bundle = authenticator.addAccount(response, "accountType", "tokenType",
      arrayOf(), mock(Bundle::class.java))

    assertNotNull(bundle)
    val intent = bundle.getParcelable<Intent>(AccountManager.KEY_INTENT)
    assertNotNull(intent)

    assertEquals(
      response,
      requireNotNull(intent).getParcelableExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE))
    assertEquals("accountType", intent?.getStringExtra(AccountManager.KEY_ACCOUNT_TYPE))
    assertEquals("tokenType", intent?.getStringExtra(AccountAuthenticator.KEY_TOKEN_TYPE))
  }

  @Test
  fun getAuthToken() {
    val response = mock(AccountAuthenticatorResponse::class.java)
    val account = Account("accountName", "accountType")
    val bundle = authenticator.getAuthToken(response, account, "tokenType", mock(Bundle::class.java))

    assertNotNull(bundle)
    val intent = bundle.getParcelable<Intent>(AccountManager.KEY_INTENT)
    assertNotNull(intent)

    assertEquals(
      response,
      requireNotNull(intent).getParcelableExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE))
    assertEquals("accountType", intent?.getStringExtra(AccountManager.KEY_ACCOUNT_TYPE))
    assertEquals("tokenType", intent?.getStringExtra(AccountAuthenticator.KEY_TOKEN_TYPE))
    assertEquals("accountName", intent?.getStringExtra(AccountManager.KEY_ACCOUNT_NAME))
  }

  @Test
  fun hasFeatures() {
    val bundle = authenticator
      .hasFeatures(mock(AccountAuthenticatorResponse::class.java), mock(Account::class.java), arrayOf())
    assertNull(bundle)
  }

  @Test
  fun updateCredentials() {
    val bundle = authenticator
      .updateCredentials(mock(AccountAuthenticatorResponse::class.java), mock(Account::class.java), "token-type",
        mock(Bundle::class.java))
    assertNull(bundle)
  }

  @Test
  fun getAuthTokenLabel() {
    val label = authenticator.getAuthTokenLabel("token-type")
    assertNull(label)
  }

  @Test
  fun editProperties() {
    val bundle = authenticator
      .editProperties(mock(AccountAuthenticatorResponse::class.java), "accountType")
    assertNull(bundle)
  }

  @Test
  fun confirmCredentials() {
    val bundle = authenticator
      .confirmCredentials(mock(AccountAuthenticatorResponse::class.java), mock(Account::class.java),
        mock(Bundle::class.java))
    assertNull(bundle)
  }
}
